### PR TITLE
Mse discontinuity

### PR DESCRIPTION
**Pull request by imbcmdth**
To support discontinuity we leverage the `timestampOffset` feature of MSE. Whenever a `timestampOffset` is set on our polyfill, (for example by contrib-hls) we force the transmuxer to reset the *BaseMediaDecodeTime* calculations so that the next segment will start at 0. When we receive that segment back from the transmuxer, we then set the `timestampOffset` of the real native source buffers to the value we were given and start appending segments.

### CODE REVIEW

File: src/transmuxer_worker.js [hunk 1]
```diff
@@ -17,39 +17,100 @@ importScripts('../node_modules/mux.js/lib/metadata-stream.js');
...
+    });
+  }
+
+  transmuxer.on('done', function (data) {
+    postMessage({ action: 'done' });
+  });
+};
 
+/**
+ * All incoming messages route through this hash. If not function exists
```
Comment (dmlap) [inline src/transmuxer_worker.js]: > If not function exists

File: src/transmuxer_worker.js [hunk 2]
```diff
@@ -17,39 +17,100 @@ importScripts('../node_modules/mux.js/lib/metadata-stream.js');
...
+    transmuxer = new muxjs.mp2t.Transmuxer(data && data.options);
+    wireTransmuxerEvents(transmuxer);
+  },
+  /**
+   * defaultInit
+   * Is called before every function and initializes the transmuxer with
+   * default options if `init` was never explicitly called
+   */
+  defaultInit: function () {
+    if (!transmuxer) {
```
Comment (dmlap) [inline src/transmuxer_worker.js]: What if this was just moved into `onmessage` below? Then we wouldn't have to remember to call it in every future message handler we create.

File: src/videojs-media-sources.js [hunk 3]
```diff
@@ -301,9 +365,9 @@
           offset += segment.byteLength;
         });
 
-        // Set timestampOffset if we have been given one
-        if (this.timestampOffset !== undefined) {
-          destinationBuffer.timestampOffset = this.timestampOffset;
+        // Set timestampOffset if it was changed before this append
```
Comment (dmlap) [inline src/videojs-media-sources.js]: Could we just set the timestampOffset on the destination buffer immediately when the property setter is invoked?

**Comment by dmlap** [comment]
We need a test to ensure that `vjs_discontinuity` is invoked on the Flash source buffer when timestampOffset is modified.

**Comment by imbcmdth** [comment]
We call vjs_discontinuity at https://github.com/videojs/videojs-contrib-media-sources/pull/39/files#diff-2e6098d381c8003336dd711b692da714R537

File: src/videojs-media-sources.js (hunk 3 above)
Comment (imbcmdth) [inline src/videojs-media-sources.js]: Maybe.

The benefit to doing it this way is that we don't have to worry about the buffers existing when `timestampOffset` is set - the offset is set only when we have done everything we needed to do to create the sourcebuffers but right before the actual `append`.

The real source buffers can be created long after the "virtual" source buffer is created. You may want to create a Media-Source object and set the `timestampOffset` before you ever start pumping data into the transmuxer (and therefore create the source buffers). This allows for that contingency without introducing too many complications.

Comment (dmlap) [inline src/videojs-media-sources.js]: Ugh. I worry this bit of asynchronicity is going to confuse us down the road. What about if we set timestampOffset directly on any buffers that are already created and make sure to check and apply the current timestamp offset as part of our lazy buffer creation process?

Comment (imbcmdth) [inline src/videojs-media-sources.js]: I think that introduces the potential for more confusion and complicates reasoning about the code.

The way it is right now, `timestampOffset` is set in exactly *one* place and that place is right before we `appendBuffer` which is exactly how a user would use `timestampOffset` in normal MSE code (ie. the `timestampOffset` is pretty much set exactly where one would expect it to be set.)

Comment (dmlap) [inline src/videojs-media-sources.js]: `timestampOffset` is side-effecting and you're shifting that side-effect in time. I don't think I'm going out on a limb to say that lazy IO is confusing. Assuming some amount of laziness is required, my proposal:

- makes the "normal" case strict
- gets rid of an extra piece of state to manage (or screw up)

Comment (dmlap) [inline src/videojs-media-sources.js]: We decided to toss the transmuxer and recreate it whenever we need to [reset the base media decode time](https://github.com/imbcmdth/videojs-contrib-media-sources/commit/1eb75f409a1ce12d25030c8a052244dff5fd7130).

**Pull request merged**